addon/components/utils/base-popover.js

```diff
@@ -0,0 +1,219 @@
+import Component from '@ember/component';
+import { guidFor } from '@ember/object/internals';
+import { run } from '@ember/runloop';
+
+import { action } from 'ember-decorators/object';
+
+import { argument, type, immutable } from 'ember-argument-decorators';
+
+import { scheduler as raf } from 'ember-raf-scheduler';
+
+function closest(node, selector) {
+  let currentNode = node;
+
+  while (currentNode !== document.body && currentNode !== null) {
+    if (currentNode.matches(selector)) {
+      return currentNode;
+    }
+
+    currentNode = currentNode.parentNode;
+  }
+
+  return null;
+}
+
+/**
+ * Base class which ice-popover, ice-tooltip, and ice-dropdown all inherit from.
+ * Adds the event listeners and data attributes that are common to all of the popover-like
+ * components, and wraps all of their functionality.
+ */
+export default class BasePopoverComponent extends Component {
```

Comment:
Yeah, I wasn't sure about naming either. I think `BasePopper` would at least separate it a little bit more, I chose not to use Popper in the name because to me it's a separate component with a clearly defined function (positioning relative to a target) but I also felt confused at times.

Some alternatives I can think of:

* BasePopMenu
* BasePopOverlay
* BasePoppable

In the end I don't think it matters too much, it being distinct from the other components is probably more important than it being distinct from Popper/AnimatedPopper 

Also, this component does not get re-exported in app, so it can't be used directly in the app by doing something like: `{{base-popover}}`, they would have to import it and re-export it. We should definitely make it clear that this is a private component though, I think we can move it to `-private` to ensure that.